### ISSUE TITLE
missing elements near the boundary

### ISSUE THREAD

**Issue by nschloe** [CGAL/cgal#5046]
A funny issue from pygalmesh (https://github.com/nschloe/pygalmesh/issues/113). When extruding a simple polygon, the resulting 
```python
import pygalmesh

points = [[0, 0], [0.1, 0.75], [0.35, 0.35], [1.0, 0.0]]

p = pygalmesh.Polygon2D(points)
domain = pygalmesh.Extrude(p, [0.0, 0.0, 1.0], 0, max_edge_size_at_feature_edges=0.1)
mesh = pygalmesh.generate_mesh(
    domain, max_cell_circumradius=0.25, max_edge_size_at_feature_edges=0.1
)
mesh.write("example_volume.vtk")
```
![screenshot](https://user-images.githubusercontent.com/181628/94837903-1b133500-0415-11eb-9fd8-18caf1e03fcf.png)

The `Extrude` class is should actually take care of the feature edges, https://github.com/nschloe/pygalmesh/blob/master/src/polygon2d.hpp#L58.

Similarly, missing elements can sometimes be observed in other examples, e.g., https://github.com/nschloe/pygalmesh/#periodic-volume-meshes.

![screenshot](https://user-images.githubusercontent.com/181628/94838322-b0aec480-0415-11eb-9d0d-beefd47182f7.png)

Is this a typical CGAL thing? Perhaps this issue in in pygalmesh. Any ideas?

**Comment by maxGimeno** [comment]
Hello, I think you should check [PMP::extrude_mesh()](https://doc.cgal.org/latest/Polygon_mesh_processing/group__PMP__meshing__grp.html#ga6cf0f4aaf1f304d129d76b9d20a4e909), it probably does what you try to do with your Extrude domain.

But to answer your question,  no, it is not typical.

**Comment by nschloe** [comment]
Mesh extrusion doesn't help me. Pygalmesh is of domain type (not mesh), meaning that you can still combine it with other domains, for example.

**Comment by sloriot** [comment]
Did you try to dumb the feature edges that are given to be protected?

**Comment by nschloe** [comment]
I did not understand that. Can you rephrase?

**Comment by sloriot** [comment]
I meant dump the edges to be protected

**Comment by MaelRL** [comment]
Concerning P3M3:

There is no missing element in P3M3's output, there is only a small dent (composed of 3 facets) in the mesh. 

The reason for this dent is that the surface is a restricted Delaunay, meaning the facets whose dual intersect the surface. Here, the dual of the fact you'd expect does not actually intersect the surface, probably not by much. Below is an illustration of what is happening, in a 2D setting (green is the input surface being meshed, grey is the mesh, blue its surface - the restricted Delaunay -, and red the dual Voronoi).

![restricted_delaunay](https://user-images.githubusercontent.com/5074170/101357214-6f191a80-3899-11eb-93ae-3a4bc0ff51ab.png)


I'm going to change the criteria value so that the surface looks more like one would expect, since it is a basic example.

**Comment by lrineau** [comment]
Should not the topological criterion `FACET_VERTICES_ON_SURFACE` fix that sort of issue?

**Comment by MaelRL** [comment]
It should. I think it is caused by the projection of dummy points, which happens post-refinement so the topology criterion might be violated then. 

It is done after refinement because I used the project_on_surface from the C3T3 helpers, but maybe it should be a projection on the mesh domain, performed ahead of the refinement phase.

**Comment by lrineau** [comment]
> It is done after refinement because I used the project_on_surface from the C3T3 helpers, but maybe it should be a projection on the mesh domain, performed ahead of the refinement phase.

For a polyhedral domain, the AABB-tree allows to project a point on the surface, but what can we do about domains based on 3D images or implicit surfaces?

**Comment by MaelRL** [comment]
P3M3 only applies to implicit surfaces right now, so on paper I can do some kind of descent.

Easier approach: as I have only a small number of points to project, I could just shoot a large number of rays from the position and take the closest result. Doesn't guarantee anything, but since it's ahead of meshing, it does not really matter either.

Optimization projections would remain the current approach, even in P3M3.

**Comment by nschloe** [comment]
Did I understand correctly that a fix for this issue is a work-in-progress right now?
